src/Microsoft.VisualStudio.Composition/Configuration/Formatters/RuntimeImportFormatter.cs

```diff
@@ -0,0 +1,132 @@
+                        : new RuntimeComposition.RuntimeImport(
+                            importingMember,
+                            importingSiteTypeRef,
+                            importingSiteTypeWithoutCollectionRef,
+                            cardinality,
+                            satisfyingExports.ToList()!,
+                            (flags & RuntimeImportFlags.IsNonSharedInstanceRequired) == RuntimeImportFlags.IsNonSharedInstanceRequired,
+                            isExportFactory,
+                            metadata,
+                            exportFactorySharingBoundaries!);
+        }
+
+        /// <inheritdoc/>
+        protected override void SerializeData(ref MessagePackWriter writer, RuntimeImport value, MessagePackSerializerOptions options)
+        {
+            RuntimeImportFlags flags = RuntimeImportFlags.None;
+            flags |= value.ImportingMemberRef == null ? RuntimeImportFlags.IsParameter : 0;
+            flags |= value.IsNonSharedInstanceRequired ? RuntimeImportFlags.IsNonSharedInstanceRequired : 0;
+            flags |= value.IsExportFactory ? RuntimeImportFlags.IsExportFactory : 0;
+            flags |=
+                value.Cardinality == ImportCardinality.ExactlyOne ? RuntimeImportFlags.CardinalityExactlyOne :
+                value.Cardinality == ImportCardinality.OneOrZero ? RuntimeImportFlags.CardinalityOneOrZero : 0;
+
+            writer.Write((byte)flags);
+
+            if (value.ImportingMemberRef is null)
+            {
+                options.Resolver.GetFormatterWithVerify<ParameterRef?>().Serialize(ref writer, value.ImportingParameterRef, options);
+            }
+            else
+            {
+                options.Resolver.GetFormatterWithVerify<MemberRef?>().Serialize(ref writer, value.ImportingMemberRef, options);
+            }
+
+            IMessagePackFormatter<TypeRef> typeRefFormatter = options.Resolver.GetFormatterWithVerify<TypeRef>();
+            typeRefFormatter.Serialize(ref writer, value.ImportingSiteTypeRef, options);
+
+            if (value.Cardinality == ImportCardinality.ZeroOrMore)
+            {
+                typeRefFormatter.Serialize(ref writer, value.ImportingSiteTypeWithoutCollectionRef, options);
+            }
+            else
+            {
+                if (value.ImportingSiteTypeWithoutCollectionRef != value.ImportingSiteTypeRef)
+                {
+                    throw new ArgumentException($"{nameof(value.ImportingSiteTypeWithoutCollectionRef)} and {nameof(value.ImportingSiteTypeRef)} must be equal when {nameof(value.Cardinality)} is not {nameof(ImportCardinality.ZeroOrMore)}.", nameof(value));
+                }
+            }
+
+            options.Resolver.GetFormatterWithVerify<IReadOnlyCollection<RuntimeExport>>().Serialize(ref writer, value.SatisfyingExports, options);
+
+            MetadataDictionaryFormatter.Instance.Serialize(ref writer, value.Metadata, options);
+            if (value.IsExportFactory)
```

Comment:
Another case of conditional serialization introducing variability into the array length.